File: client/src/game/ClientBot.ts
/**
 * Client-side bot that takes over flipper and launcher control.
 *
 * Pure logic — no PixiJS, DOM, or physics engine dependencies.
 * Game.ts feeds ball state in, gets flipper/launcher commands out.
 *
 * Coordinates: physics meters (PPM=500). Y-axis positive = down.
 * Flipper Y ≈ 1.16m, board center X ≈ 0.4m.
 */

/** Minimal ball state needed by the bot. */
export interface BallInfo {
  x: number;
  y: number;
  vx: number;
  vy: number;
  inLauncher: boolean;
  inShooterLane: boolean;
}

/** Commands returned by the bot each tick. */
export interface BotOutput {
  leftFlipper: boolean;
  rightFlipper: boolean;
  launch: boolean;
}

// Flip zone: only react when ball is below this Y (in meters)
const FLIP_ZONE_Y = 1.0;
// Board center X in meters (left half → left flipper, right half → right)
const CENTER_X = 0.4;
// Minimum time between flips per flipper (seconds)
const FLIP_COOLDOWN = 0.2;
// How long to hold a flip (seconds)
const FLIP_HOLD = 0.15;

// Launcher: how long to charge before releasing (seconds)
const LAUNCH_CHARGE_MIN = 0.3;
const LAUNCH_CHARGE_MAX = 0.7;
// Cooldown after a launch before launching again
const LAUNCH_COOLDOWN = 1.0;

export class ClientBot {
  private leftHold = 0;
  private rightHold = 0;
  private leftCooldown = 0;
  private rightCooldown = 0;

  private launchTarget = 0; // how long to charge this launch
  private launchHeld = 0; // how long we've been holding
  private launching = false;
  private launchCooldown = 0;

  /** Seed for simple deterministic pseudo-random variation. */
  private seed = 1;

  private nextRandom(): number {
    // Simple LCG for light variation (no crypto needed)
    this.seed = (this.seed * 1664525 + 1013904223) & 0x7fffffff;
    return this.seed / 0x7fffffff;
  }

  update(dt: number, balls: BallInfo[]): BotOutput {
    // Decrement timers
    this.leftHold = Math.max(0, this.leftHold - dt);
    this.rightHold = Math.max(0, this.rightHold - dt);
    this.leftCooldown = Math.max(0, this.leftCooldown - dt);
    this.rightCooldown = Math.max(0, this.rightCooldown - dt);
    this.launchCooldown = Math.max(0, this.launchCooldown - dt);

    // --- Flipper logic ---
    // Find the lowest active ball (highest y) that's in the flip zone and moving down
    let flipBall: BallInfo | null = null;
    for (const b of balls) {
      if (b.inLauncher || b.inShooterLane) continue;
      if (b.y < FLIP_ZONE_Y) continue;
      if (b.vy <= 0) continue; // moving up, ignore
      if (!flipBall || b.y > flipBall.y) {
        flipBall = b;
      }
    }

    if (flipBall) {
      if (flipBall.x < CENTER_X && this.leftCooldown <= 0) {
        this.leftHold = FLIP_HOLD;
        this.leftCooldown = FLIP_COOLDOWN;
      }
      if (flipBall.x >= CENTER_X && this.rightCooldown <= 0) {
        this.rightHold = FLIP_HOLD;
        this.rightCooldown = FLIP_COOLDOWN;
      }
    }

    // --- Launcher logic ---
    const hasLaunchBall = balls.some((b) => b.inLauncher || b.inShooterLane);
    let launch = false;

    if (hasLaunchBall && this.launchCooldown <= 0) {
      if (!this.launching) {
        // Start a new launch: pick a random charge duration
        this.launching = true;
        this.launchHeld = 0;
        this.launchTarget =
          LAUNCH_CHARGE_MIN +
          this.nextRandom() * (LAUNCH_CHARGE_MAX - LAUNCH_CHARGE_MIN);
      }

      this.launchHeld += dt;

      if (this.launchHeld < this.launchTarget) {
        // Still charging: hold the button
        launch = true;
      } else {
        // Release: launcher fires on this edge (true→false handled by next tick)
        launch = false;
        this.launching = false;
        this.launchCooldown = LAUNCH_COOLDOWN;
      }
    } else if (this.launching && !hasLaunchBall) {
      // Ball left the launcher area (was launched or escaped)
      this.launching = false;
    }

    return {
      leftFlipper: this.leftHold > 0,
      rightFlipper: this.rightHold > 0,
      launch,
    };
  }

  /** Reset all state (called when bot is toggled off). */
  reset(): void {
    this.leftHold = 0;
    this.rightHold = 0;
    this.leftCooldown = 0;
    this.rightCooldown = 0;
    this.launchTarget = 0;
    this.launchHeld = 0;
    this.launching = false;
    this.launchCooldown = 0;
  }
}

File: client_bevy/src/game/client_bot.rs
/// Client-side bot that takes over flipper and launcher control.
/// Mirrors the TypeScript bot behavior to keep gameplay feel consistent.

#[derive(Clone, Copy, Debug)]
pub(crate) struct BotBallInfo {
    pub(crate) x: f32,
    pub(crate) y: f32,
    pub(crate) vx: f32,
    pub(crate) vy: f32,
    pub(crate) in_launcher: bool,
    pub(crate) in_shooter_lane: bool,
}

#[derive(Clone, Copy, Debug, Default, PartialEq)]
pub(crate) struct BotOutput {
    pub(crate) left_flipper: bool,
    pub(crate) right_flipper: bool,
    pub(crate) launch: bool,
}

const FLIP_ZONE_Y: f32 = 0.85;
const CENTER_X: f32 = 0.4;
const FLIP_COOLDOWN: f32 = 0.6;
const FLIP_HOLD: f32 = 0.2;
const MIN_BALL_SPEED: f32 = 0.3;

const LAUNCH_CHARGE_MIN: f32 = 0.7;
const LAUNCH_CHARGE_MAX: f32 = 1.0;
const LAUNCH_COOLDOWN: f32 = 1.5;

#[derive(Debug)]
pub(crate) struct ClientBot {
    left_hold: f32,
    right_hold: f32,
    left_cooldown: f32,
    right_cooldown: f32,
    launch_target: f32,
    launch_held: f32,
    launching: bool,
    launch_cooldown: f32,
    seed: u32,
}

impl Default for ClientBot {
    fn default() -> Self {
        Self {
            left_hold: 0.0,
            right_hold: 0.0,
            left_cooldown: 0.0,
            right_cooldown: 0.0,
            launch_target: 0.0,
            launch_held: 0.0,
            launching: false,
            launch_cooldown: 0.0,
            seed: 1,
        }
    }
}

impl ClientBot {
    fn next_random(&mut self) -> f32 {
        self.seed = self.seed.wrapping_mul(1664525).wrapping_add(1013904223) & 0x7fff_ffff;
        self.seed as f32 / 0x7fff_ffffu32 as f32
    }

    pub(crate) fn update(&mut self, dt: f32, balls: &[BotBallInfo]) -> BotOutput {
        self.left_hold = (self.left_hold - dt).max(0.0);
        self.right_hold = (self.right_hold - dt).max(0.0);
        self.left_cooldown = (self.left_cooldown - dt).max(0.0);
        self.right_cooldown = (self.right_cooldown - dt).max(0.0);
        self.launch_cooldown = (self.launch_cooldown - dt).max(0.0);

        let mut flip_ball: Option<&BotBallInfo> = None;
        for ball in balls {
            if ball.in_launcher || ball.in_shooter_lane {
                continue;
            }
            if ball.y < FLIP_ZONE_Y {
                continue;
            }
            let speed = (ball.vx * ball.vx + ball.vy * ball.vy).sqrt();
            if speed < MIN_BALL_SPEED {
                continue;
            }
            if flip_ball.is_none_or(|current| ball.y > current.y) {
                flip_ball = Some(ball);
            }
        }

        if let Some(ball) = flip_ball {
            let left_side = ball.x < CENTER_X - 0.05;
            let right_side = ball.x > CENTER_X + 0.05;
            if !left_side && !right_side {
                if self.left_cooldown <= 0.0 {
                    self.left_hold = FLIP_HOLD;
                    self.left_cooldown = FLIP_COOLDOWN;
                }
                if self.right_cooldown <= 0.0 {
                    self.right_hold = FLIP_HOLD;
                    self.right_cooldown = FLIP_COOLDOWN;
                }
            } else if left_side && self.left_cooldown <= 0.0 {
                self.left_hold = FLIP_HOLD;
                self.left_cooldown = FLIP_COOLDOWN;
            } else if right_side && self.right_cooldown <= 0.0 {
                self.right_hold = FLIP_HOLD;
                self.right_cooldown = FLIP_COOLDOWN;
            }
        }

        let has_launch_ball = balls.iter().any(|b| b.in_launcher || b.in_shooter_lane);
        let mut launch = false;
        if has_launch_ball && self.launch_cooldown <= 0.0 {
            if !self.launching {
                self.launching = true;
                self.launch_held = 0.0;
                self.launch_target = LAUNCH_CHARGE_MIN
                    + self.next_random() * (LAUNCH_CHARGE_MAX - LAUNCH_CHARGE_MIN);
            }

            self.launch_held += dt;
            if self.launch_held < self.launch_target {
                launch = true;
            } else {
                launch = false;
                self.launching = false;
                self.launch_cooldown = LAUNCH_COOLDOWN;
            }
        } else if self.launching && !has_launch_ball {
            self.launching = false;
        }

        BotOutput {
            left_flipper: self.left_hold > 0.0,
            right_flipper: self.right_hold > 0.0,
            launch,
        }
    }

    pub(crate) fn reset(&mut self) {
        self.left_hold = 0.0;
        self.right_hold = 0.0;
        self.left_cooldown = 0.0;
        self.right_cooldown = 0.0;
        self.launch_target = 0.0;
        self.launch_held = 0.0;
        self.launching = false;
        self.launch_cooldown = 0.0;
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn bot_flips_for_ball_in_zone() {
        let mut bot = ClientBot::default();
        let out = bot.update(
            1.0 / 60.0,
            &[BotBallInfo {
                x: 0.2,
                y: 1.0,
                vx: 0.5,
                vy: 0.5,
                in_launcher: false,
                in_shooter_lane: false,
            }],
        );
        assert!(out.left_flipper || out.right_flipper);
    }

    #[test]
    fn bot_holds_launch_while_charging() {
        let mut bot = ClientBot::default();
        let balls = [BotBallInfo {
            x: 0.0,
            y: 0.0,
            vx: 0.0,
            vy: 0.0,
            in_launcher: true,
            in_shooter_lane: true,
        }];
        let out = bot.update(1.0 / 60.0, &balls);
        assert!(out.launch);
    }
}
